Add certificate based authentication in check_sync and pldap/pldaps support

davidcoutadeur opened the pull request:
Original PR propose by @damiencassu, see #59 for discussions about these features

Closes #63
Close #64

damiencassu commented:
Hello @davidcoutadeur , 

I've tested this new version on my test environment, as it it doesn't work, I had to made minor adjustments (I will point them in a code review)

damiencassu commented on slapd-cli.conf:
```diff
@@ -7,9 +7,15 @@ IP="*"
 LDAPI_SOCKETURL="%2Fvar%2Frun%2Fslapd%2Fldapi"
 SLAPD_SERVICES="ldap://$IP:$PORT ldaps://$SSLIP:$SSLPORT ldapi://$LDAPI_SOCKETURL"
+# For enabling HAProxy proxy protocol
+#SLAPD_SERVICES="pldap://$PIP:$PPORT pldaps://$PSSLIP:$PSSLPORT ldapi://$LDAPI_SOCKETURL"
```
Maybe make this example clearer ?
Here you can be confused and comment the default SLAPD_SERVICES variable to only uncomment the second one. Thus you end up disabling ldapi and standard ldap endpoints which can cause issues with the script and with slapd replication. 

Maybe something like below would be more suitable ?
`# For enabling HAProxy proxy protocol, add pldap and/or pldaps endpoint as below`

`#SLAPD_SERVICES="ldap://$IP:$PORT ldaps://$SSLIP:$SSLPORT ldapi://$LDAPI_SOCKETURL pldap://$PIP:$PPORT pldaps://$PSSLIP:$PSSLPORT `

davidcoutadeur replied:
should be ok with last commit

davidcoutadeur commented on slapd-cli:
```diff
@@ -1365,11 +1372,28 @@ check_sync() {
+				else
+					CERTVARS='LDAPTLS_CACERT=${PCACERT} LDAPTLS_CERT=${PCERT} LDAPTLS_KEY=${PKEY}'
+				fi
+				CONTEXTCSN=$(eval ${CERTVARS} && ${LDAPSEARCH_BIN} -Y EXTERNAL -H "${PHOST}" -b "${PBASEDN}" -o nettimeout=5 -s base ${LDAPSEARCH_PARAMS} contextCSN -LLL 2>/dev/null )
```
Could you give more details about the malformed ldapsearch? Which part?

I have tested the && part, it seems that the variables are correctly evaluated and passed to the second command.

damiencassu replied:
Yes sure !

With your original ldapsearch the script can't perform the request and fails. When I add -d 1 to the ldapsearch command I get the following details:

ldap_sasl_interactive_bind: Can't contact LDAP server (-1)
        additional info: error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed (self signed certificate in certificate chain)

It seems the CERTVARS are not taken into account. I've always used ldpasearch with certs variables as a oneliner (without && or without defining them in a previous command). 

Just copy-pasting the generated command in my terminal fails while without the "&&" it works.

davidcoutadeur replied:
Strange... On my side it works only with &&.

Could you try with the last commit I just pushed. It surrounds the vars and the eval with double quotes: a8ecf15

If you still have the error, could you try to run slapd-cli with `bash -x`?

davidcoutadeur replied:
I have made a full test. There was indeed a problem with &&, as the variables were not passed.
I have also added the LDAPTLS_REQCERT="try" variable. It should be ok now.
@damiencassu could you make a final test to check if everything is ok for you?

davidcoutadeur replied:
The assumption is that when tls_cert and tls_key are set, we try TLS mutual authentication. We may have get the parameter from olcSyncrepl, but in a first time, it seems simpler to just have this static parameter

davidcoutadeur commented on slapd-cli:
```diff
@@ -1344,6 +1350,7 @@ check_sync() {
 
 	LOCAL_URL="${SLAPD_SERVICES% }"
 	LOCAL_URL="$( printf ${LOCAL_URL} | sed -e 's/\*/localhost/g' )"
+	LOCAL_URL_SSL=$( printf "%s" "${SLAPD_SERVICES}" | sed -e 's/\*/localhost/g' -e 's/^.*\(ldaps:[^ \t\r\n]\+\).*$/\1/' )
```
Thank you for the report, indeed, a fix was needed

davidcoutadeur commented:
I have squashed all commits, and made a last cleaning task.

I have one last remark: I have removed `LDAPTLS_REQCERT=never` in simple bind commands. (in checksync function)

Indeed, if you have configured your master provider with ldaps, the right solution is to configure TLS_REQCERT and TLS_CACERT in your ldap.conf or .ldaprc. Those files are read by the ldapsearch commands called in slapd-cli

davidcoutadeur commented:
This PR seems all good to me. I am merging.
@damiencassu : if you have time to test, don't hesitate to give your feedback

damiencassu commented:
Hello @davidcoutadeur, 

Sorry I did not have time to test until this morning.

I agree for the `LDAPTLS_REQCERT=never` it should have been removed by previous commits...strange !

Your lastest version also works (even with the "try" while my test configuration is set up with "demand").

Seems you forgot to change the script version from 3.1 to 3.2 :)

Thanks a lot for taking the time to review and support this PR, have a nice week !

Regards,

davidcoutadeur commented:
> Hello @davidcoutadeur,
> 
> Sorry I did not have time to test until this morning.
> 
> I agree for the `LDAPTLS_REQCERT=never` it should have been removed by previous commits...strange !
> 
> Your lastest version also works (even with the "try" while my test configuration is set up with "demand").
> 
> Seems you forgot to change the script version from 3.1 to 3.2 :)
> 
> Thanks a lot for taking the time to review and support this PR, have a nice week !
> 
> Regards,

Thanks for your last feedback!

"try" should work as well as "demand" in this case.

The script version will be updated for imminent 3.3 release.

Thanks for your contribution!
